### DISCUSSION TITLE
Shared Objects/Libraries within Fprime Framework

### DISCUSSION THREAD

**Discussion by ebenezerarunkumar** [nasa/fprime#1224] [Q&A]
Hi! 

I am currently working on a project where we have received a recommendation that the software executable we deliver be split up into shared objects/libraries to limit the scope of required patching in flight.

Is this possible with Fprime? I haven't seen anyone do anything like that. The reason why this might be useful for our mission is that we are incorporating a GNC library that was written by a different collaborator into our Fprime framework, and it would be useful to have the functionality to *only* patch the GNC library on-orbit instead of replacing the binary file.

How has JPL approached patching software on its Fprime missions?

**Comment by timcanham** [comment]
What is the operating system on your target hardware?

**Reply by ebenezerarunkumar** [reply]
It is running a Linux OS

**Reply by timcanham** [reply]
I would defer to @LeStarch, but I think it is just a cmake option to generate shared libraries instead of a static link.

**Reply by LeStarch** [reply]
fprime-util generate -DBUILD_SHARED_LIBS=ON should do it. Note:  shared libs can dredge up dependency issues.  Our CMake UTs do cover this case, but since it isn't frequently used by customers you may run into friction.

Please comment back if trouble arises.

https://cmake.org/cmake/help/latest/variable
/BUILD_SHARED_LIBS.html

**Reply by LeStarch** [reply]
Also note, if you are using shared libs to limit updated scope **be careful** as changes to the ABI of the library will break upgrades.

Why not consider a binary patch process instead?

**Comment by Joshua-Anderson** [comment]
#1178 is a similar discussion on this topic.

If your primary concern is limiting filesize of updates one approach to consider is binary diff patches.

You can make changes to the flight software and rebuild, then create a binary diff between the onboard version and the new release.
This is often simpler than trying to break flight software into independently updatable modules while minimizing the amount of data that has to be uplinked for the update.

**Comment by ebenezerarunkumar** [comment]
Thank you all for the input! I agree that a binary diff approach would probably be the easier option. However, I am bound by the requirements of the project I am on, and they have asked us to try to use shared libraries. 

But anyway, I will make sure to keep you guys in the loop with how using shared libraries goes, if we decide to use that approach!

Thanks!
